README.md
## BrukerEC_PS moved to https://gitlab.com/srgblnch-tangocs/BrukerEC_PS


NEWS
1.12.2
* improve robustness of download verification feature to avoid trouble while
  import BO correctors
* tuning.py: eliminates trace at startup caused by DevFailed (RegulationTuneable attribute not found)
* cabinet.py restart(): call start() with keyword argument re
* BrukerBend: only tries to set_write_value() for WRITE-able attributes
* BrukerBend: made WaveName READ only (both because TANGO C++ bug 3068607
      and writing is not sensible anyway unless both bendings use the same ramp)
* adds forgotten 0x0E and 0x0F states to dipole / quadrupole power supplies
* setpoint readback outside regular limits (after restart)_ will not affect reading of CurrentSetpoint

1.12.1
* make difference in WaveName be okay (green, VALID) instead of alarm (orange)
* cabinet.py:checked_command() fix for wrong status code being returned
* cabinet.py:checked_command() fix for harmless can bus timeout blocking
  communication
* slightly changing cabinet status message for cabinets without relay board
* increased socket timeout for reads to 2.5, after seeing socket.timeouts
  where the power supply is working fine
* remove Waveform from unified interface
* socket.error will now cause a (transitory) communication error message to be
  shown (short termn communication errors that appear and disapper quickly
  should NOT be observed during normal operation)

1.12.0
* not pushing WaveName anymore working around a bug in Tango fixed in 7.2
* added ErrorCode2,3,4 for bend / quadrupole power supplies
* ErrorCode includes now the state as well
* removed command UpdateState from Tango level to avoid accidentically polling it
* major internal changes in cabinet state handling, in order to provide
  explanation of ErrorCode attributes

1.11
* fixes problem preventing regulation parameters to be written even when
  RegulationTuneable was set to True
* ResetInterlocks: limits calls to bsw_tcp_repeater2 to at most once every 10 s
* lowers command read timeout to 2 seconds
* replaces push_changing by push_vdq
* removed all_initialized check
* eliminates bogus ControlTrace at init when cab.stat was None
* fixes some pychecker warnings related to imports
* using BUSY for waveform uploads and downloads
* PSType.mask_cab replaced by Cabinet.mask
* customize cabinet interlock messages
* retain telnet connections

1.10
* queueing writes while loading waveforms
* fixes telnet commands related hangs
* improved of connection refused errors (cPCI being abruptly shutdown)
* removed all_initialized check

1.9
* improved synchronization feature added to BrukerBend_PS.py
* error messages for slave and buck power supply modules

Communication
* improved handling of control units going offline, and coming back
* retry can bus commands upto three times
* can bus errors will only block the originating power supply, so others will
  remain usable
* wave up- and downloads will not cause can bus errors anymore


wavl.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# wavl.py
# TANGO Device Server (http://sourceforge.net/projects/tango-ds/)
#
# Copyright (c) 2009 by None
#
# GNU General Public Licence (GPL)
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 3 of the License, or (at your option) any later
# version.
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc., 59 Temple
# Place, Suite 330, Boston, MA  02111-1307  USA

'''This module contains functionality related uploading and downloading
waveforms for the Bruker power supplies from ALBA.
In includes mainly WaveformLoader class and two Thread classes managing
the upload respective download of a waveform.
'''

from __future__ import with_statement

# imports from Standard Library
import os, os.path
from time import sleep, time
from threading import Lock, Thread
from collections import deque
from types import NoneType
import numpy as numpy
from zlib import crc32
from functools import partial

# extra modules
import PowerSupply.util as PU
import PowerSupply.standard as PS

# local imports
from wavl_hash import *

WAVE_PORT = 3702
TERM = '\r\n'
MAX_WAVE_LEN = 16384
AVG_TIME_PER_SAMPLE = 0.0013*3
# minimal time required for download plus safety margin
TIME_BASE = 0.45+1.0
# waveform points can have signed integer values from 0x80000 up to 0x7ffff
# value corresponding to nominal setting
PT_MAX = 0x7ffff
PT_MIN = -PT_MAX

READY = 'ready'

_WAVL = None #< recent most created wave form loader

def instance():
    global _WAVL
    if not _WAVL:
        _WAVL = WaveformLoader()
    return _WAVL

def dump_to_file(self, wup, wdown, attempt):
        '''never raises an Exception
        '''
        try:
            dirname = os.path.join(self.dumpdir, str(os.getpid()))
            try:
                os.makedirs(dirname)
            except OSError, err:
                self.log.debug('making new dump dir failed: ' + str(err))
            fname_up = os.path.join(dirname, str(attempt)+'-up.txt')
            fname_down = os.path.join(dirname, str(attempt)+'-down.txt')
            open(fname_up, 'w').write('\n'.join(map(repr,wup)))
            open(fname_down, 'w').write('\n'.join(map(repr,wdown)))
            self.log.info('dumped waveforms into {0}'.format(dirname))
        except Exception, exc:
            self.log.warn(
                'verification failed AND ' +
                'exception while dumping waveforms into {0}'.format(dirname),
                exc_info=1)
            return exc

class Load(object):
    def __init__(self, port):
        self.port = port

    __repr__ = lambda self: self.__str__


class Download(Load):

    BASE_MSG = 'download'

    def __init__(self, port, wave, data):
        Load.__init__(self, port)
        self.wave = wave
        self.data = data

    def __str__(self):
      return 'Download(port %d, %d pt)' % (self.port, len(self.data))

    def run(self, impl, wavl):
        was_power_on = impl.is_power_on()
        impl.ramp_off()
        if impl.is_ramping():#it would have to have change to 'off' with ramp_off() call
            raise PS.PS_Exception('Trying to download a waveform when ramping')
        rval = wavl.download(self.port, self.data)
        impl.push_wave_down(self.wave)
        if was_power_on: impl.On()
        return rval

class Upload(Load):

    BASE_MSG = 'upload'

    def __init__(self, port, maxlen=None):
        Load.__init__(self, port=port)
        self.maxlen = maxlen

    def __str__(self):
      return 'Upload(port %d, %s pt)' % (self.port, self.maxlen)

    def run(self, impl, wavl):
      rval = wavl.upload(self.port, maxlen=self.maxlen)
      impl.push_wave_up(self, rval)
      return rval

class WaveformException(PS.PS_Exception):

    def __init__(self, response):
        self.response = response
        self.code = int(response[2:4])
        text = response[5:]
        PS.PS_Exception.__init__(self, text)

class LoadException(PS.PS_Exception):
    pass

def bebusy(msg):

    def wrapper_maker(f):
        def wrap_busy(self, *args, **kwargs):

            self.busy = msg
            try:
                return f(self, *args,**kwargs)
            finally:
                self.busy = None
        return wrap_busy

    return wrapper_maker

class WaveformLoader(object):
    '''Managing the thread for up- and downloading waveforms.
    '''

    active_load = None
    busy = None

    # how many times verification shall be attempted before generating an error
    verify = 3

    def __init__(self):
        global _WAVL
        _WAVL = self
        self.log = None
        self.sok = PU.FriendlySocket()
        self.sok.reconnect_delay = 20
        self.sok.connect_timeout = 2.0
        # up- and downloads set their own timeouts
        self.sok.read_timeout = TIME_BASE
        self.dumpdir = '/tmp/BrukerEC-wavl/'

    is_connected = property(lambda self: self.sok.is_connected)

    def connect(self, host):
        self.sok.connect(host, WAVE_PORT)

    def disconnect_exc(self):
        self.sok.disconnect_exc()

    def reconnect(self):
        self.sok.reconnect()

    def __del__(self):
        self.disconnect_exc()

    @bebusy('uploading ramp')
    def upload(self, ch, maxlen=None):
        '''reads waveform from hardware, discarding the last point.
           since the server appends a copy of the first point to then end.
        '''
        sok = self.sok
        if not sok.is_connected:
            raise PS.PS_Exception('not connected')
        cmd_start_ul = 'u'+str(ch)+TERM
        wave = deque()
        tmout = AVG_TIME_PER_SAMPLE*maxlen + TIME_BASE
        self.log.debug('starting upload %s, timeout %s', ch, tmout)
        start_t = time()
        self.sok.write(cmd_start_ul)
        if maxlen is None: maxlen = MAX_WAVE_LEN
        self.sok.timeout = tmout
        try_again = True
        while True:
            payload = self.sok.readline().strip()
            if payload.startswith('ok'):
                break

            elif payload == 'er08' and try_again:
                self.log.debug('was busy, trying again')
                try_again = False
                self.sok.write(cmd_start_ul)
                continue

            elif payload.startswith('e'):
                raise WaveformException(payload)

            elif payload == '':
                self.sok.disconnect_exc()
                raise PS.PS_Exception('disconnected from wave loader')
            else:
                try:
                    wave.append(int(payload))
                except Exception:
                    raise PS.PS_Exception('strange payload %r in upload, read %d points so far' % (payload, len(wave)))

        self._duration = time()-start_t
        wave = tuple(wave)
        if maxlen and maxlen<len(wave):
            msg = 'upload yielded %d points, expected maximal %d' % (len(wave), maxlen)
            raise PS.PS_Exception(msg)

        self.log.debug('upload time %s s for %s pt', self._duration, len(wave))
        return wave


    def check_waveform(self, dat):
        """Generates exception when trying to transmit values that are out of
           valid range, which is given by PT_MAX (two times the nominal value).
        """
        max_dat = max(dat)
        if max_dat > PT_MAX:
            idx = dat.index(max_dat)
            msg = 'waveform point %d too big: %d > %d' % (idx, max_dat, PT_MAX)
            raise PS.PS_Exception(msg)

        min_dat = min(dat)
        if min_dat < PT_MIN:
            idx = dat.index(min_dat)
            msg = 'waveform point %d too small: %d < %d' % (idx, min_dat, PT_MIN)
            raise PS.PS_Exception(msg)

    @bebusy('downloading ramp')
    def download(self, ch, wave):
        '''Transmits waveform to control unit.
        '''
        if not self.sok.is_connected:
            raise PS.PS_Exception('not connected')

        wave = tuple(int(w) for w in wave)
        tmout = self.sok.timeout
        self.log.debug('starting download %s, timeout %s s', ch, tmout)
        self.sok.timeout = self.sok.read_timeout
        start_dl_cmd = 'd'+str(ch)+TERM
        self.sok.write(start_dl_cmd)
        buf = ''.join(str(w)+TERM for w in wave) + ';' + TERM
        self.log.debug('buf contains %d characters' % len(buf))
        start_t = time()
        self.sok.timeout = AVG_TIME_PER_SAMPLE * len(wave) + TIME_BASE
        self.sok.write(buf)
        ret = self.sok.readline().strip()
        if not ret.startswith('ok'):
            raise WaveformException(ret)
        duration = time()-start_t
        self.log.info('download {0} finished after {1} s'.format(ch, duration))
        if self.verify:
            self.log.debug('verifying...')
            success = False
        else:
            success = True

        for r in range(1,1+self.verify):
            upstart_t = time()
            wup = self.upload(ch,len(wave))
            up_duration = time()-upstart_t
            if tuple(wup)==tuple(wave):
                self.log.info('verification succeeded at {0}. attempt'.format(r))
                success = True
                break
            else:
                self.log.warn('verification failed, attempt {0}'.format(r))
                dump_to_file(self, wup, wave, r)

        if not success:
            raise LoadException('verification failed')
        self._duration = duration
        return ret

    @bebusy('cancel')
    def cancel(self):
        self.sok.write('!'+TERM)
        r = self.sok.readline().strip()
        self.log.debug('cancel: %s',r)
        return r



tuning.py
# -*- coding: utf-8 -*-

# py standard library
from time import sleep
import traceback

# TANGO imports
import PyTango as Tg
import PowerSupply.standard as PS

PSEUDO_ATTR = PS.PseudoAttr()

class Tuner(object):

    writeable = False
    attr_list = None

    def add_RegulationTuneable(self):

        @PS.AttrExc
        def read_RegulationTuneable(inst, attr):
            attr.set_value(self.writeable)

        @PS.AttrExc
        def write_RegulationTuneable(inst, attr):
            self.writeable = attr.get_write_value()
            inst.push_change_event('RegulationTuneable', self.writeable)

        attr = Tg.Attr('RegulationTuneable', Tg.DevBoolean, Tg.READ_WRITE)
        self.impl.add_attribute(attr,
            r_meth=read_RegulationTuneable,
            w_meth=write_RegulationTuneable)
        while True:
            try:
                self.impl.set_change_event('RegulationTuneable', True)
                break
            except ValueError:
                self.impl.warn('add_RegulationTuneable', exc_info=1)
                sleep(0.1)
            except Tg.DevFailed, df:
                if df[-1].reason=='API_AttrNotFound':
                    self.impl.warn('add_RegulationTuneable', exc_info=1)
                    sleep(0.1)
                else:
                    raise



        aprop = Tg.UserDefaultAttrProp()
        aprop.set_description('whether modifying parameters related to regulation loop should be allowed or not. only set this to true when you really, really know what you are doing AND have a backup.')
        attr.set_default_properties(aprop)

    def __init__(self, impl, pstype):
        self.impl = impl
        self.pstype = pstype
        self.add_RegulationTuneable()
        # the tuneable flag is held in write_value of RegulationTuneable attribute
        self.RegulationTuneable = impl._attr('RegulationTuneable')
        self.RegulationTuneable.set_write_value(False)

        # Regulation Loop Control
        self.has_buck = len(pstype.module)>1
        master_slave = len(pstype.module)>2

        self.attr_list = []
        self.add_ports()
        if master_slave:
            self.add_ports(slave=True)

    def add_attr(self, aname, rp, slave_index, **extra):
        impl = self.impl
        extra['format'] = '%7.4f'
        index = rp.index+slave_index
        @PS.AttrExc
        def read_f(inst, attr):
            inst.obj_vdq(rp.cobj, index).set_attr(attr)

        @PS.AttrExc
        def write_f(inst, wattr):
            if not self.writeable:
                raise PS.PS_Exception('tuning regulation parameters disabled')
            write_value = wattr.get_write_value()
            inst.obj_set(rp.cobj, index, repr(write_value))

        attr = Tg.Attr(aname, Tg.DevDouble, rp.rw)
        aprop = Tg.UserDefaultAttrProp()

        if 'desc' in extra:
            extra['description'] = extra['desc']
            del extra['desc']

        attr.set_disp_level(Tg.DispLevel.EXPERT)

        for key,val in extra.iteritems():
            getattr(aprop, 'set_'+key)(val)

        # retries for 10 seconds...
        for x in range(0,40):
            try:
                self.impl.add_attribute(attr, r_meth=read_f, w_meth=write_f)
                break
            except Exception, exc:
                self.impl.log.debug(str(exc), exc_info=1)
                sleep(0.25)

        attr.set_default_properties(aprop)
        self.attr_list.append(attr)

        # sets a useful write value
        if rp.rw == Tg.READ_WRITE:
            try:
                read_value = impl.obj_vdq(rp.cobj, index).value
                write_attr = impl._attr(aname)
                write_attr.set_write_value(read_value)
                impl.log.info('attribute %r added %r',aname, read_value)
            except Tg.DevFailed, fail:
                err = fail[-1]
                impl.log.info('attribute %r added but failed reading (%s)', aname, err.desc)

    def remove_all(self):
        for attr in self.attr_list:
            try:
                self.impl.remove_attribute(attr.get_name())
            except ValueError,exc:
                self.impl.log.debug(attr.get_name()+': '+str(exc))

    def add_ports(self, slave=False):
        # changing regulation parameters is disabled
        # as it is normally not done and inappropriate values might damage the power supply
        pre = 'Slave' if slave else ''
        slave_index = 2 if slave else 0
        for aname, rp in self.pstype.REG_PARAM.iteritems():
            self.add_attr(pre+aname, rp, slave_index)


wavl_hash.py
#!/usr/bin/python

from zlib import crc32
import numpy

# maximal value used in internal representation of waveform (18 bits)
PT_NOMINAL =  0x3ffff 

# module shared between client and server
# used to generate wave id

def to_raw(I_nominal, waveform):
    '''converts waveform (floats) into integer representation used
       by loading waveforms to Bruker power supply
       @param I_nominal nominal current of power supply
       @param waveform input waveform
    '''
    return tuple( int( (y*PT_NOMINAL)/I_nominal ) for y in waveform )

def calc_hash(I_nominal, waveform):
    '''Calculates hash value for the specified waveform.
       @param I nominal current of the power supply
       @param waveform input waveform
    '''
    dat = (int(I_nominal),) + to_raw(I_nominal, waveform)
    ar = numpy.array(dat, numpy.int32)
    return crc32(ar.data)

def hex_u32(value):
    """Converts a hash code into string of unsigned hexadecimal digits.       
       -1 will be 2*31 result in fffffffff
    """
    if value < 0:
        value += 2**32
    return "%x" % value

def unhex_i32(string):
  """Converts string of (unsigned) hexadecimal digits back into signed 32 bit integer
  """
  val = int(string,16)
  if val >=2**31:
      val -= 2**32
  return val 

def fingerprint(I_nominal, waveform):
    '''Returns fingerprint (str) that should be shown to the user.
       @param I_nominal nominal current of power supply
       @param waveform input waveform
    '''
    hash = calc_hash(I_nominal, waveform)
    return hex_u32(hash)

if __name__ == '__main__':
    for x in (-1, -2**31, 0, 2**31-1, 3):
	    print x, hex_u32(x), unhex_i32(hex_u32(x))
    import sys
    if len(sys.argv) < 3:
        print 'usage: ', sys.argv[0], ' I_nominal FILE_NAME'
        sys.exit(1)
    I = float(sys.argv[1])
    wav_fname = sys.argv[2]
    wav = numpy.loadtxt(wav_fname)
    print fingerprint(I, wav)



state.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# state.py
# TANGO Device Server (http://sourceforge.net/projects/tango-ds/)

'''Bruker power supply state and status handling.
'''

# Standard Library
from copy import copy, deepcopy

# Extra Packages
from PyTango import DevState
import PyTango as Tg

import PowerSupply.standard as PS
from PowerSupply.util import bitpack2, bitunpack, bit_filter_msg

PM = PS.MSG

TYPE_1 = {
    0x00: (DevState.ALARM, 'GSM_CONFIG'),
    0x01: (DevState.ALARM, 'GSM_CONFIG_1' ),
    0x02: (DevState.ALARM, 'SYNC' ),
    0x03: (DevState.OFF, 'IDLE: relay opened, pulse off (DC OFF)' ),
    0x04: (DevState.FAULT, 'ADC_CAL'),
    0x05: (DevState.INIT, 'INRUSH'),
    0x06: (DevState.ON, 'DCON: power on' ),
    0x07: (DevState.STANDBY, 'STANDBY' ),
    0x08: (DevState.INIT, 'PULSE ON'),
    0x09: (DevState.ALARM, 'REG ON'),
    0x0a: (DevState.ON, 'RUNNING: relay closed, pulse and regulation on (DC ON)'),
    0x0b: (DevState.INIT, 'STOPPING'),
    0x0c: (DevState.ALARM, 'FAULT: relay opened, pulse and regulation off, error pending'),
    0x0d: (DevState.ALARM, 'FAULT_ZC'),
}

TYPE_3 = deepcopy(TYPE_1)
TYPE_3.update({
    0x00: (DevState.INIT, 'GSM_CONFIG: waiting for synchronization command'),
    0x00: (DevState.INIT, 'GSM_CONFIG_1'),
    0x03: (DevState.OFF, 'IDLE: pulse off (DC OFF)' ),
    0x08: (DevState.INIT, 'PULSEON: activating 4Q PWM and regulation'),
    0x09: (DevState.INIT, 'REGON: send DC ON order, waiting weak board(s) to reach RUNNING state'),
    0x0a: (DevState.ON, 'RUNNING: pulse and regulation on (DC ON)'),
    0x0c: (DevState.ALARM, 'FAULT_PN: pulse and regulation on, error pending'),
    0x0d: (DevState.ALARM, 'FAULT_PF: pulse and regulation off, error pending'),
    0x0e: (DevState.INIT, 'STOPPING_1' ),
    0x0f: (DevState.INIT, 'STOPPING_2' ),
})

TYPE_4 = deepcopy(TYPE_1)
TYPE_6 = deepcopy(TYPE_1)
TYPE_7 = deepcopy(TYPE_1)
TYPE_8 = deepcopy(TYPE_3)
TYPE_8.update({
    0x03: (DevState.OFF, 'IDLE: relay opened, pulse off (DC OFF)' ),
})

class Module(object):
    """Represents a regulation board
    """
    error_code = None
    errors = None
    machine_stat = {
        0 : (DevState.UNKNOWN, 'not applicable'),
        None : (DevState.UNKNOWN, 'not applicable'),
    }
    alias = ()
    fullname = None


    def __init__(self, name, errors=None, machine_stat=None, fullname=None):
        if not name is None:
            self.name = name
        if not errors is None:
            self.errors = errors
        if not machine_stat is None:
            self.machine_stat = machine_stat
        if fullname is None:
            self.fullname = self.name
        else:
            self.fullname = fullname

        # required for update_stat2
        self.alarms = []
        self.state_id = None

    def __str__(self):
        return '%s' % self.name

    __repr__ = __str__

    def update_stat2(self):
        '''Updates stat attribute from self.state_id and self.error_code
        '''
        self.alarms = bit_filter_msg(self.error_code, self.errors)
        if not self.errors:
            self.stat = DevState.INIT, 'module type not detected (yet)'
        elif self.alarms:
            status = '\n'.join(self.alarms)
            self.stat = DevState.ALARM, status
        elif self.state_id is None:
            self.stat = [ DevState.UNKNOWN, 'okay' ]
        elif self.state_id > len(self.machine_stat):
            self.stat = [ DevState.FAULT, 'unknown state [%02d]' % self.state_id ]
        else:
            self.stat = self.machine_stat[self.state_id]


class RP(object):
      """Description of 1 RegulationParameter."""
      def __init__(self, cobj, index=0, w=1):
          self.cobj = cobj
          self.index = index
          self.rw = Tg.READ_WRITE if w else Tg.READ


class PSType(object):
    '''Description of the different types of power supply
       in particular interpretation of error flags.
    '''

    has_trigger_mask = False

    REG_PARAM = {
        'BuckV' : RP(0x2041, w=0),
        'BuckV_Kp' : RP(0x2055),
        'BuckV_Ki' : RP(0x2053),
        'MainI_Kp' : RP(0x2026),
        'MainI_Ki' : RP(0x2024),
        'MainV_Kp' : RP(0x2029),
        'MainV_Ki' : RP(0x2027)
    }

    def __init__(self, name, *module_list, **kwargs):
        self.name = name
        self.module = module_list
        self.XI = kwargs.get('XI', ())
        # if not on well, it is off
        self.states_on = (10, )
        self.states_switch_off = kwargs.get('swoff', (11, 14, 15) )
        self.states_switch_on = kwargs.get('swon', (7,8,9 ) )
        self.states_standby = kwargs.get('standby', () )
        self.use_waveforms = True

    def update_xi(self, msg_list):
        '''Updates the message displayed for the external interlocks of this
           power supply.
        '''
        p0 = self.module[0]
        for idx,msg in zip(self.XI,msg_list):
            p0.errors[idx] = msg

    def __str__(self):
        return self.name + " PS "+ str(self.module)

    def __repr__(self):
        return self.__class__.__name__ + repr( (self.name, self.module) )

    def query_Voltage(self, impl):
        cmd = impl.cab.command
        master = impl.Port
        cmd_adv = lambda p: float(cmd(p, 'ADV/'))
        adv = cmd_adv(master)
        return PS.VDQ(adv, q=PS.AQ_VALID)


class PSType_SmallQuad(PSType):

      def __init__(self, name, *ports, **kwargs):
          PSType.__init__(self, name, *ports, **kwargs)
          if not 'XI' in kwargs:
              self.XI = range(12,16)

      has_trigger_mask = True

      REG_PARAM = {
        'BuckV' : RP(0x200B,1, w=0),
        'BuckI_Kp' : RP(0x2026,1),
        'BuckI_Ki' : RP(0x2024,1),
        'BuckV_Kp' : RP(0x2029,1),
        'BuckV_Ki' : RP(0x2027,1),

        'MainI_Kp' : RP(0x2026,0),
        'MainI_Ki' : RP(0x2024,0),
        'MainV_Kp' : RP(0x2029,0),
        'MainV_Ki' : RP(0x2027,0),
      }

      def query_Voltage(self, impl):
        cmd = impl.cab.command
        master = impl.Port+1
        cmd_adv = lambda p: float(cmd(p, 'ADV/'))
        adv = cmd_adv(master)
        return PS.VDQ(adv, q=PS.AQ_VALID)

class PSType_Big(PSType_SmallQuad):

      def query_Voltage(self, impl):
        cmd = impl.cab.command
        cmd_adv = lambda p: float(cmd(p, 'ADV/'))
        adv_master = cmd_adv(impl.Port+0)
        adv_slave = cmd_adv(impl.Port+2)
        return PS.VDQ(adv_master+adv_slave, q=PS.AQ_VALID)

# common error messages
EBIT = lambda x: PM.UNDOCUMENTED_BIT.format('error', x)
E_EARTH = 'earth leakage'
E_EEPROM = 'EEPROM error'
E_BIG_LOAD = 'load over-current, voltage or RMS current'
E_LOAD = 'load over-current or over-voltage'
E_ADC = 'ADC voltage or watchdog'
E_DCCT = 'DCCT fault'
E_BUCK_TEMP = 'buck overtemperature'
E_EXTERN = [ 'extern 1', 'extern 2', 'extern 3', 'extern 4' ]
E_CAPV = 'capacitor voltage unbalanced'
E_IGBT = 'IGBT fault'

# the integers are used as symbolic constants
# errors for t1 (correctors and sextupoles)
ERRORS_CORR = [
    ## Error byte, LSB, 0x00 ... 0x08
    E_EARTH,
    'DC filter overvoltage (or {0})'.format(E_EEPROM),
    E_LOAD,
    E_ADC,
    'DC on fault',
    EBIT(5),
    E_DCCT,
    E_IGBT,

    ## Error byte, MSB, 0x00 ... 0x08
    'power stage overtemperature',
    'switcher overtemperature',
    'rectifier overtemperature',
    'transformer overtemperature'
] + E_EXTERN
assert(len(ERRORS_CORR)==16)

# errors for t3 (quad 4Q)
ERRORS_QUAD = [
    ## Error byte, LSB, 0x00 ... 0x80
    E_EARTH,
    E_EEPROM,
    E_BIG_LOAD,
    E_ADC,
    E_DCCT,
    E_CAPV,
    'switcher overtemperature',
    '4Q IGBT fault (bit 7)',

    ## Error byte, MSB, 0x0100 ... 0x8000
    'unbalance',
    '4Q IGBT fault (bit 9)',
    E_BUCK_TEMP,
    EBIT(11),
] + E_EXTERN
assert(len(ERRORS_QUAD)==16)

# errors for t8 (bending 4Q)
ERRORS_BEND = [
    ## Error byte, LSB, 0x00 ... 0x80
    E_EARTH,
    E_EEPROM,
    E_BIG_LOAD,
    E_ADC,
    E_DCCT,
    'IGBT 2Q2 (2) fault',
    E_BUCK_TEMP,
    'oscillation',

    ## Error byte, MSB, 0x0100 ... 0x8000
    E_CAPV,
    'IGBT 2Q1 (2) fault',
    'IGBT 2Q2 (1) fault',
    'IGBT 2Q1 (2) fault',
    'unbalance 2Q2 (2)',
    'self 2 overtemperature',
    'unbalance 2Q1 (1)',
    'self 1 overtemperature',
]

# errors for t4 (buck)
ERRORS_BUCK = [
    'buck over voltage',
    'EEPROM error',
    'buck over current',
    'ADC U/watchdog',
    'rectifier over voltage',
    EBIT(5),
    EBIT(6),
    EBIT(7),

## MSB
    EBIT(8),
    EBIT(9),
    EBIT(10),
    E_IGBT,
    EBIT(12),
    EBIT(13),
    EBIT(14),
    EBIT(15),
]
assert(len(ERRORS_BUCK)==16)

# global errors
ERRORS_CABINET = [
    'cabinet CAN communication',
    'cabinet EPROM',
    EBIT(2),
    'cabinet watch dog',
    'communication cabinet',
    'phase',
    'door open',
    'water',
# MSB
    EBIT(8),
    'cabinet fuse',
    'emergency / PSS',
    'temperature cabinet'
] + E_EXTERN
assert(len(ERRORS_CABINET)==16)

MOD_C = Module('corrector', ERRORS_CORR, TYPE_1)
MOD_C.alias = ('cor', 'c', 't1')
MOD_Q = Module('4Q', ERRORS_QUAD, TYPE_3, 'quadrupole 4Q standalone')
MOD_Q.alias = ( 'q', 't3', 'quad', 'quad standalone')
MOD_QM = Module('4Q master', ERRORS_BEND, TYPE_3, 'quadrupole 4Q master' )
MOD_QM.alias = ('qm', 'quad master')
MOD_QS = Module('4Q slave', ERRORS_BEND, TYPE_3, 'quadrupole 4Q slave' )
MOD_QS.alias = ('qs', 'quad slave')
MOD_BUCK = Module('buck', ERRORS_BUCK, TYPE_4 )
MOD_BUCK.alias = ('t4', )
MOD_BUCK1 = Module('buck 1', ERRORS_BUCK, TYPE_4 )
MOD_BUCK1.alias = ('buck1',)
MOD_BUCK2 = Module('buck 2', ERRORS_BUCK, TYPE_4 )
MOD_BUCK2.alias = ('buck2',)
MOD_S = Module('sextupole', ERRORS_CORR, TYPE_7 )
MOD_S.alias = ('sex', '6', 't7')
MOD_BM = Module('4Q master', ERRORS_BEND, TYPE_8, 'bending 4Q master' )
MOD_BM.alias = ( 'bend', 'bend master', 'bm', 't8')
MOD_BS = Module('4Q slave', ERRORS_BEND, TYPE_8, 'bending 4Q slave' )
MOD_BS.alias = ( 'bend', 'bend slave', 'bs')
CABINET_XI = xrange(12, 16)

# quadrupole types
PSTYPE_CODE_QUAD = 3

# defining PsType objects
# quad with 2 sub-modules, ( QS120/180, QC320/340 )
PSTYPE_SMALL_QUAD = PSType_SmallQuad('small quadrupole',
  MOD_Q, MOD_BUCK
)
# quad with 4 sub-modules ( QC340 )
PSTYPE_BIG_QUAD = PSType_Big('big quadrupole',
  MOD_QM, MOD_BUCK1, MOD_QS, MOD_BUCK2
)

PSTYPE_BEND = PSType_Big('bending',
  MOD_BM, MOD_BUCK1, MOD_BS, MOD_BUCK2, Isafe=470)

# sextupole type
PSTYPE_SEX = PSType('sextupole', MOD_C, XI=xrange(12, 16))
PSTYPE_SEX.REG_PARAM = copy(PSType.REG_PARAM)
del PSTYPE_SEX.REG_PARAM['BuckV']

# maps software types to PSTypes
REG2PSTYPE = {
    1 : PSType('corrector', MOD_C, XI=xrange(12, 16) ), # correctors and quadrupoles?
    2 : PSType('LT bend', MOD_C, XI=xrange(12, 16)), # bending magnets of the LT cabinet
    3 : None, # is a quadrupole but could have 2 or 4 submodules, so further work required
    4 : None, #< buck for ???
    5 : None, # quadrupole relay board
    6 : PSType('BT', MOD_C, XI=xrange(12,16)), #< code used by both buck and BT,
    7 : PSTYPE_SEX,
    8 : PSTYPE_BEND,
    9 :  None, #< dipole relay board
}


# interpretation of cobj_BRKconfig 2 object
BRKconfig2_MASTER = 0
BRKconfig2_STANDALONES = (1,2)
BRKconfig2_SLAVE = 3

COBJ_MAIN_REGUL_FREQ = 0x200A
COBJ_MAIN_IREF_SCALED = 0x2020
COBJ_MAIN_IREF_LIMIT = 0x2021
COBJ_VERSION = 0x2034
COBJ_BRK_CONFIG2 = 0x2038

class __ModuleRegistry(object):


    def __init__(self):
        self.name_dict = dict()
        self.name_list = []
        self.module_list = []

    def register(self, *modules):
        for m in modules:
            self.module_list.append(m)
            self.name_dict[m.fullname.lower()] = m

            for a in m.alias:
                self.name_dict[a] = m

    def get(self, name):
        mod = self.name_dict.get(name.lower())
        return mod

    def ls1(self, m):
        names = [ m.fullname ]
        names +=  m.alias
        return names

    def ls(self):
        return [ ', '.join(self.ls1(m)) for m in self.module_list ]

MODULE_REGISTRY = __ModuleRegistry()
MODULE_REGISTRY.register(MOD_C, MOD_Q, MOD_QM, MOD_QS, MOD_BUCK, MOD_BUCK1,
    MOD_BUCK2, MOD_S, MOD_BM, MOD_BS)

# dummy object that is never equal to a 'real' state (integer)
STATE_NONE = object()

BITCOUNT = (7,24)
BITOVER = (0,0)

def synthesize_error_code(st, error_code):
    code, overflow = bitpack2(BITCOUNT, BITOVER, (st, error_code) )
    assert not overflow[0], 'state code {0} too big'.format(st)
    assert not overflow[1], 'error code {0} too big'.format(code)
    return code

def analyze_error_code(code):
    return bitunpack(BITCOUNT, BITOVER, code)[0]

def format_module_stat(mod):
    alarm_text = ''
    code = synthesize_error_code(mod.state_id, mod.error_code)
    if mod.alarms:
        for a in mod.alarms:
            alarm_text += '\n    ' + a
    else:
        alarm_text = 'no alarm'
    try:
        status = mod.machine_stat[mod.state_id][1]
    except (KeyError,IndexError):
        status = 'undocumented state code'
    return (
        'board: {0}\n'.format(mod.fullname)+
        'code: {0:#x}\n'.format(code)+
        'status: {0:#x} {1}\n'.format(mod.state_id, status) +
        'error code: {0:#x}\n'.format(mod.error_code) +
        'alarms: ' +  alarm_text
    )

BOARD_LIST = [
    MOD_C,
    MOD_Q,
    MOD_BUCK,
    MOD_S ,
]




